### PR TITLE
Add option to selectively restore model weights and optimizer states in AutoResume and MegatronStrategy

### PR DESCRIPTION
**Pull request by hemildesai**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/lightning/pytorch/strategies.py [hunk 1]
```diff
@@ -628,17 +637,39 @@ def load_checkpoint(self, checkpoint_path: Union[str, Path]) -> Dict[str, Any]:
         sharded_state_dict = {}
         sharded_state_dict["state_dict"] = self.megatron_parallel.sharded_state_dict()
 
-        if self.ckpt_include_optimizer and self.trainer.state.fn == TrainerFn.FITTING:
+        if (self.ckpt_include_optimizer or include_optimizer_states) and self.trainer.state.fn == TrainerFn.FITTING:
```
Comment (ashors1) [inline nemo/lightning/pytorch/strategies.py:640]: I feel like it's a bit confusing to have two different arguments with nearly the same name. What about instead we add `ckpt_load_optimizer` and `ckpt_save_optimizer` attributes to the strategy?

File: nemo/lightning/resume.py [hunk 2]
```diff
@@ -1,107 +1,81 @@
...
+    resume_past_end: bool = False
+    resume_ignore_no_checkpoint: bool = False
+    restore_optimizer_states: bool = False
 
     def setup(self, trainer: Union[pl.Trainer, fl.Fabric], model=None):
         if isinstance(trainer, fl.Fabric):
             raise NotImplementedError("Fabric is not supported yet.")
 
-        ckpt_path = self.nemo_path(model)
+        ckpt_path, restore_path = self.nemo_path(model)
```
Comment (ashors1) [inline nemo/lightning/resume.py:60]: could we rename these to make it clear that one path is for the optimizer and the other is for the model?

File: nemo/lightning/pytorch/strategies.py (hunk 1 above)
Comment (hemildesai) [reply nemo/lightning/pytorch/strategies.py:640]: Yeah I think that should work. I'm ok with that.

File: nemo/lightning/resume.py (hunk 2 above)
Comment (hemildesai) [reply nemo/lightning/resume.py:60]: The `ckpt_path` is for `trainer.ckpt_path` whereas the `restore_path` is for selective restore. I can rename `ckpt_path` to `trainer_ckpt_path`.

File: nemo/lightning/pytorch/strategies.py (hunk 1 above)
Comment (ashors1) [reply nemo/lightning/pytorch/strategies.py:640]: Good point. What about just renaming `include_optimizer_states` to `load_optimizer_states_from_ckpt` or something similar?

File: nemo/lightning/resume.py [hunk 3]
```diff
@@ -1,107 +1,81 @@
...
+    resume_past_end: bool = False
+    resume_ignore_no_checkpoint: bool = False
+    restore_optimizer_states: bool = False
 
     def setup(self, trainer: Union[pl.Trainer, fl.Fabric], model=None):
         if isinstance(trainer, fl.Fabric):
             raise NotImplementedError("Fabric is not supported yet.")
 
-        ckpt_path = self.nemo_path(model)
+        ckpt_path, selective_restore_path = self.nemo_path(model)
```
Comment (yaoyu-33) [inline nemo/lightning/resume.py:60]: can we add a comment explain what are the two returns here?

File: nemo/lightning/resume.py [hunk 4]
```diff
@@ -148,18 +140,52 @@ def nemo_path(self, model=None) -> Optional[Path]:
...
+                    if base_model_path != Path(metadata['model_ckpt_path']):
+                        raise ValueError(
+                            f"When trying to resume a PEFT training run, found mismatching values: "
+                            f"your specified restore_path points to {base_model_path}, "
+                            f"but the PEFT checkpoint was trained with "
+                            f"model_ckpt_path={metadata['model_ckpt_path']}"
+                        )
+
+                    # return AdapterPath(base_model_path, adapter_path=checkpoint), None
+                    return AdapterPath(checkpoint, base_model_path=base_model_path), None
```
Comment (hemildesai) [inline nemo/lightning/resume.py:172]: Can you move this to a separate private method?

**Review by hemildesai — COMMENTED** [review]
@cuichenx Two small comments for the PEFT changes

File: nemo/lightning/resume.py [hunk 5]
```diff
@@ -1,107 +1,97 @@
...
+            restore_path = model.import_ckpt(self.restore_path)
+        except (ValueError, AttributeError):
+            restore_path = self.restore_path
 
-class AutoResume(Resume, io.IOMixin):
-    """Class that handles the logic for setting checkpoint paths and restoring from
-    checkpoints in NeMo.
-    """
+        if self.adapter_path:
+            restore_path = AdapterPath(Path(self.adapter_path), adapter_path=restore_path)
```
[the remark this answers is not in the document]
Comment (cuichenx) [reply nemo/lightning/resume.py:77]: yes sorry, missed this

File: nemo/lightning/resume.py (hunk 4 above)
Comment (cuichenx) [reply nemo/lightning/resume.py:172]: good idea

File: nemo/collections/llm/gpt/data/fine_tuning.py [hunk 6]
```diff
@@ -54,10 +54,7 @@ def __init__(
         self.seq_length = seq_length
         self.seed = seed
         self.dataset_root = Path(dataset_root)
-
-        from nemo.collections.nlp.modules.common.tokenizer_utils import get_nmt_tokenizer
-
-        self.tokenizer = tokenizer or get_nmt_tokenizer("megatron", "GPT2BPETokenizer")
+        self.tokenizer = tokenizer
```
Comment (ashors1) [inline nemo/collections/llm/gpt/data/fine_tuning.py:57]: What's the reason for this change?

Comment (cuichenx) [reply nemo/collections/llm/gpt/data/fine_tuning.py:57]: I don't think it made sense to put a default tokenizer here. Can you tell me your use case? In practice, it was downloading a megatron-345m model every time, so I wanted to prevent that

File: tests/lightning/test_nemo_logger.py [hunk 7]
```diff
@@ -82,21 +82,21 @@ def test_resume(self, trainer, tmp_path):
         # Error because explicit_log_dir does not exist
         with pytest.raises(NotFoundError):
             nl.AutoResume(
-                dirpath=str(tmp_path / "test_resume"),
+                resume_from_directory=str(tmp_path / "test_resume"),
```
Comment (yaoyu-33) [inline tests/lightning/test_nemo_logger.py:85]: why do I only see `resume_from_directory` change in this test file?

File: nemo/collections/llm/recipes/llama3_8b.py [hunk 8]
```diff
@@ -96,7 +95,10 @@ def pretrain_recipe(
 
 
 def hf_resume() -> Config[nl.AutoResume]:
-    return Config(nl.AutoResume, import_path="hf://meta-llama/Meta-Llama-3.1-8B")
+    return Config(
+        nl.AutoResume,
+        selective_restore_config=Config(nl.SelectiveRestoreConfig, path="hf://meta-llama/Meta-Llama-3.1-8B"),
```
Comment (cuichenx) [inline nemo/collections/llm/recipes/llama3_8b.py:100]: This didn't originate from this PR, but this hf path should be `hf://meta-llama/Meta-Llama-3-8B` to match the `Llama3Config8B` above. (Alternatively change above to `Llama31Config8B`, and same for other recipe files.

Comment (cuichenx) [reply nemo/collections/llm/recipes/llama3_8b.py:100]: let me know if you think this is okay, i can also make the change

Comment (hemildesai) [reply nemo/collections/llm/recipes/llama3_8b.py:100]: Oh yeah feel free to change, I must've missed this.

File: nemo/collections/llm/gpt/data/fine_tuning.py (hunk 6 above)
Comment (ashors1) [reply nemo/collections/llm/gpt/data/fine_tuning.py:57]: Makes sense, was just wondering why this was added to this PR. The tokenizer is no longer optional now, right? Should we change the type hint too?

File: nemo/lightning/pytorch/strategies/utils.py [hunk 9]
```diff
@@ -21,6 +22,14 @@
 from nemo.utils.callbacks.dist_ckpt_io import AsyncFinalizableCheckpointIO
 
 
+@dataclass(kw_only=True)
+class SelectiveRestoreConfig:
```
Comment (akoumpa) [inline nemo/lightning/pytorch/strategies/utils.py:26]: `RestoreConfig` seems like a better option, "Config" conveys the selective part by definition.

File: nemo/lightning/pytorch/strategies/utils.py [hunk 10]
```diff
@@ -21,6 +22,14 @@
...
+    path: str
+    adapter_path: Optional[str] = None
+    model_weights: bool = True
```
Comment (akoumpa) [inline nemo/lightning/pytorch/strategies/utils.py:29]: How about `load_model_state` & `load_optim_state`?

I want to avoid a verb here but also make sure the options aren't confused with the actual thing.

File: nemo/collections/llm/recipes/mistral.py [hunk 11]
```diff
@@ -32,7 +32,7 @@ def trainer(devices=8) -> nl.Trainer:
 
 @factory(name=NAME + "_hf")
 def hf_resume() -> nl.AutoResume:
-    return nl.AutoResume(import_path="hf://mistralai/Mistral-7B-v0.3")
+    return nl.AutoResume(selective_restore_config=nl.SelectiveRestoreConfig(path="hf://mistralai/Mistral-7B-v0.3"))
```
Comment (akoumpa) [inline nemo/collections/llm/recipes/mistral.py:35]: why isn't this `Config(nl.SeclectiveRestoreConfig ...` ?

File: nemo/lightning/pytorch/strategies/megatron_strategy.py [hunk 12]
```diff
@@ -602,13 +611,19 @@ def save_checkpoint(
...
             checkpoint["sharded_state_dict"] = self.megatron_parallel.sharded_state_dict()
-        if self.trainer.state.fn == TrainerFn.FITTING and self.ckpt_include_optimizer:
+        if self.trainer.state.fn == TrainerFn.FITTING and self.ckpt_save_optimizer:
             checkpoint["optimizer"] = [self.optimizer_sharded_state_dict()]
 
         self.checkpoint_io.save_checkpoint(checkpoint, filepath, storage_options=storage_options)
 
+    def should_restore_optimizer_states(self, selective_restore: bool = False) -> bool:
+        if selective_restore:
+            return self.selective_restore_config.optimizer_states if self.selective_restore_config else False
```
Comment (akoumpa) [inline nemo/lightning/pytorch/strategies/megatron_strategy.py:621]: Would that cause any trouble? We are using m argument and a class attribute's value to make a decision. 

Is there any way to make this only use the argument?

File: nemo/lightning/resume.py [hunk 13]
```diff
@@ -1,183 +1,209 @@
...
     BasePath = WindowsPath
 else:
     BasePath = PosixPath
 
 
-class Resume(IOMixin):
-    def nemo_path(self, model=None) -> Optional[Path]:
-        """Returns the checkpoint to resume from."""
+@dataclass(kw_only=True)
+class AutoResume:
```
Comment (akoumpa) [inline nemo/lightning/resume.py:24]: we need to rename this class, but I'm not sure it should happen in this PR.

File: nemo/lightning/resume.py [hunk 14]
```diff
@@ -1,183 +1,209 @@
...
+            trainer.strategy.selective_restore_config = self.selective_restore_config
+
+    def _try_import_model(
+        self, model: Optional[io.ConnectorMixin], path: str, adapter_path: Optional[str] = None
+    ) -> BasePath:
+        if model is None:
+            raise ValueError("Model is needed to import checkpoint from HF or other non-NeMo checkpoint format.")
+        try:
+            new_path = model.import_ckpt(path)
+        except (ValueError, AttributeError):
```
Comment (akoumpa) [inline nemo/lightning/resume.py:83]: what triggers the ValueError / AttributeError here?

File: nemo/lightning/resume.py [hunk 15]
```diff
@@ -1,183 +1,209 @@
...
-                    checkpoint = last_checkpoints[0]
-                    checkpoint = uninject_model_parallel_rank(checkpoint)
-                else:
-                    # Select the checkpoint with the latest modified time
-                    checkpoint = sorted(last_checkpoints, key=lambda pth: pth.lstat().st_mtime, reverse=True)[0]
-                    logging.warning(
-                        f"Multiple checkpoints {last_checkpoints} matches *last.ckpt. Selecting one with the latest modified time."
-                    )
+        elif len(end_checkpoints) > 0:
+            if self.resume_past_end:
```
Comment (akoumpa) [inline nemo/lightning/resume.py:149]: make this `if not self.resume_past_end`, raise the exception and move the nested block one level up.

File: nemo/lightning/resume.py [hunk 16]
```diff
@@ -1,183 +1,209 @@
...
+            True and a checkpoint matching ``*end.ckpt`` indicating a previous training run
+            fully completed. Setting resume_past_end=True disables this behavior and loads the
+            last checkpoint.
+        resume_ignore_no_checkpoint (bool): AutoResume throws an error if resume_if_exists is
+            True and no checkpoint could be found. Setting resume_ignore_no_checkpoint=True
+            disables this behavior, in which case exp_manager will print a message and
+            continue without restoring.
+    """
+
+    selective_restore_config: Optional[SelectiveRestoreConfig] = None
```
Comment (akoumpa) [inline nemo/lightning/resume.py:48]: Would it be possible to rename this `restore_config` (prev. comment) and also allow passing a string, in which case a `SelectiveRestoreConfig` is created in the constructor and restores everything the checkpoint contains?

Comment (hemildesai) [reply nemo/lightning/resume.py:48]: I had named it `resotre_` before but @ashors1 suggested to use `selective_restore` to make it more explicit. I don't have a strong preference for either, but I don't think being explicit hurts in this case.

File: nemo/lightning/resume.py (hunk 14 above)
Comment (hemildesai) [reply nemo/lightning/resume.py:83]: It's for when the connector is not found for the particular path.

File: nemo/lightning/resume.py [hunk 17]
```diff
@@ -1,183 +1,209 @@
...
-        self.resume_if_exists = resume_if_exists
-        self.resume_past_end = resume_past_end
-        self.resume_ignore_no_checkpoint = resume_ignore_no_checkpoint
-
-    def nemo_path(self, model=None) -> Optional[Path]:
+        trainer_ckpt_path = self.get_trainer_ckpt_path(model)
+        if trainer_ckpt_path:
+            trainer.ckpt_path = trainer_ckpt_path
+            trainer.checkpoint_callback.last_model_path = trainer_ckpt_path
+        elif self.selective_restore_config:
```
Comment (akoumpa) [inline nemo/lightning/resume.py:63]: I think there's a bug here, my suggestion is to change `elif` to `if`.

**Review by akoumpa — CHANGES_REQUESTED** [review]

File: nemo/lightning/resume.py (hunk 17 above)
Comment (hemildesai) [reply nemo/lightning/resume.py:63]: This is intended as you cannot have both together. Basically, you want to skip selective restore if checkpoint path is found.

File: nemo/lightning/pytorch/strategies/megatron_strategy.py (hunk 12 above)
Comment (hemildesai) [reply nemo/lightning/pytorch/strategies/megatron_strategy.py:621]: It's not possible to only use the argument since the helper is called by methods like `load_checkpoint` which are invoked by internal PTL APIs without any args. The arg `selective_restore` just tells the helper whether you're in the selective restore stage or not. Open to suggestions, but couldn't find a better way to accomplish this.

File: nemo/lightning/pytorch/strategies/utils.py (hunk 10 above)
Comment (hemildesai) [reply nemo/lightning/pytorch/strategies/utils.py:29]: Yeah I'm fine with the rename, will do.

File: nemo/lightning/resume.py (hunk 17 above)
Comment (hemildesai) [reply nemo/lightning/resume.py:63]: Raising an error would not support the following usecase:
1. Start finetuning from SelectiveRestoreConfig
2. Resume finetuning from the last saved checkpoint by 1.

File: nemo/lightning/resume.py (hunk 14 above)
Comment (akoumpa) [reply nemo/lightning/resume.py:83]: @hemildesai thanks, can you add include your GH comment as a code comment?

File: nemo/lightning/pytorch/strategies/utils.py (hunk 10 above)
Comment (akoumpa) [reply nemo/lightning/pytorch/strategies/utils.py:29]: thank you.

File: nemo/lightning/resume.py (hunk 16 above)
Comment (akoumpa) [reply nemo/lightning/resume.py:48]: I don't see the value of `Selective` here.

Comment (hemildesai) [reply nemo/lightning/resume.py:48]: @ashors1 @yaoyu-33 Are you fine with just `RestoreConfig`?

File: nemo/collections/llm/gpt/data/fine_tuning.py (hunk 6 above)
Comment (cuichenx) [reply nemo/collections/llm/gpt/data/fine_tuning.py:57]: A user doesn't need to specify this, because the llm.finetune API will link the tokenizer from the imported model

File: nemo/collections/llm/recipes/mistral.py (hunk 11 above)
Comment (hemildesai) [reply nemo/collections/llm/recipes/mistral.py:35]: Looks like the recipes in this file are not updated. MIght require a separate PR to update all recipes here.

File: nemo/lightning/resume.py (hunk 16 above)
Comment (hemildesai) [reply nemo/lightning/resume.py]: Updated to `RestoreConfig` in https://github.com/NVIDIA/NeMo/pull/10295/commits/f240efad57edf08aeaa2d9b2bfc73724b865ca6d

File: nemo/collections/llm/recipes/llama3_8b.py (hunk 8 above)
Comment (hemildesai) [reply nemo/collections/llm/recipes/llama3_8b.py]: Updated.

**Review by akoumpa — DISMISSED** [review]
LGTM, thanks.

**Review by akoumpa — APPROVED** [review]
